components/layout/display_list_builder.rs

```diff
@@ -1414,22 +1415,56 @@ impl FragmentDisplayListBuilding for Fragment {
                                                   gradient: &Gradient,
                                                   style: &ComputedValues,
                                                   index: usize) {
+        // Calculate where the first "tile" needs to be placed on one axis.
+        // * base is the beginning of the visible area
+        // * start is the current "tile" position
+        // * tile is the length of the "tile"
+        // Returns a difference between start and new start.
+        // It holds that base - tile < start - diff <= base
+        fn get_first_tile(base: i32, start: i32, tile: i32) -> i32 {
+            if tile == 0 {
+                return 0;
+            }
+            if start > base {
+                ((start - base) / tile + 1) * tile
+            } else {
+                ((base - start) / tile) * tile
+            }
+        }
+
         let bg = style.get_background();
+        let bg_origin = get_cyclic(&bg.background_origin.0, index).clone();
         let bg_size = get_cyclic(&bg.background_size.0, index).clone();
         let bg_position_x = get_cyclic(&bg.background_position_x.0, index).clone();
         let bg_position_y = get_cyclic(&bg.background_position_y.0, index).clone();
-        let border = self.border_width().to_physical(style.writing_mode);
 
-        let mut bounds = *absolute_bounds;
-        bounds.origin.x = bounds.origin.x + border.left + bg_position_x.to_used_value(bounds.size.width);
-        bounds.origin.y = bounds.origin.y + border.top + bg_position_y.to_used_value(bounds.size.height);
-        bounds.size.width = bounds.size.width - border.horizontal();
-        bounds.size.height = bounds.size.height - border.vertical();
+        let mut bounds = absolute_bounds;
+
+        match bg_origin {
+            BackgroundOrigin::BorderBox => {}
+            BackgroundOrigin::PaddingBox => {
+                let border = style.logical_border_width().to_physical(style.writing_mode);
+                bounds.origin.x = bounds.origin.x + border.left;
+                bounds.origin.y = bounds.origin.y + border.top;
+                bounds.size.width = bounds.size.width - border.horizontal();
+                bounds.size.height = bounds.size.height - border.vertical();
+            }
+            BackgroundOrigin::ContentBox => {
+                let border_padding = self.border_padding.to_physical(style.writing_mode);
+                bounds.origin.x = bounds.origin.x + border_padding.left;
+                bounds.origin.y = bounds.origin.y + border_padding.top;
+                bounds.size.width = bounds.size.width - border_padding.horizontal();
+                bounds.size.height = bounds.size.height - border_padding.vertical();
+            }
+        }
+
+        bounds.origin.x = bounds.origin.x + bg_position_x.to_used_value(bounds.size.width);
+        bounds.origin.y = bounds.origin.y + bg_position_y.to_used_value(bounds.size.height);
```

Comment:
Could you change all the above code to use `+=` and `-=` operators, instead of `a = a + b`?

I know the previous code used the long form, but I find the `+=` form much easier to read and review.